### DISCUSSION TITLE
Possible inclusion of a timezone list

### DISCUSSION THREAD

**Discussion by PHYSXGH** [khoih-prog/Timezone_Generic#8] [Ideas]
Hi all,

I was wondering if there exists a list of all timezones around the world with the DST rules defined for this library, as I'm working on a clock and was planning to use this library, but manually defining all of the TimeChangeRules and Timezones would take quite a bit of time. On the other hand, if there isn't anything like that, I might add my definitions to the repo once they are complete, if anyone would like to use them.

Thanks

**Comment by khoih-prog** [comment]
HI @PHYSXGH 

That's possibly good idea and easy to do, but will take a hard work and much time for any library maintainer, because the DST rules are changing quickly, and also complex.

It's better that this is the work of **interested users**, which possess **more knowledge and time** to manage a separate file for these rules.

Just create an `.h` file then include it.

For example

https://github.com/khoih-prog/Timezone_Generic/blob/367b3dd3e7aca6c7c7f2e65d40e517a311856d34/examples/RP2040W/TZ_NTP_WorldClock_RP2040W/TZ_NTP_WorldClock_RP2040W.ino#L24-L69

**Comment by khoih-prog** [comment]
Hi @PHYSXGH 

Just remember I used a TZ list in [**ESP_WiFiManager TZ.h**](https://github.com/khoih-prog/ESP_WiFiManager/blob/master/src/utils/TZ.h) or [**ESPAsync_WiFiManager TZ.h**](https://github.com/khoih-prog/ESPAsync_WiFiManager/tree/master/src/utils), which can be auto-generated. You can try to use this way, if possible, to save lot of your precious time.

```
// autogenerated from https://raw.githubusercontent.com/nayarsystems/posix_tz_db/master/zones.csv
// by script <esp8266 arduino core>/tools/TZupdate.sh
// Thu Nov 12 04:07:03 UTC 2020
//
// This database is autogenerated from IANA timezone database
//    https://raw.githubusercontent.com/nayarsystems/posix_tz_db/master/zones.csv
//    (using https://www.iana.org/time-zones)
// and can be updated on demand in this repository
// or by yourself using the above script

//////////////////////////////////////////

//See: https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html
// EST5EDT,M3.2.0,M11.1.0 (for America/New_York)
// EST5EDT is the name of the time zone
// EST is the abbreviation used when DST is off
// 6 hours is the time difference from GMT
// EDT is the abbreviation used when DST is on
// ,M3 is the third month
// .2 is the second occurrence of the day in the month
// .0 is Sunday
// ,M11 is the eleventh month
// .1 is the first occurrence of the day in the month
// .0 is Sunday
```

**Comment by PHYSXGH** [comment]
Thanks, that's really useful. I've been thinking about the issue over the past couple days and my use case is essentially a clock which allows the user to select the timezone using a web interface. Built on a Pico W. The main point is that it shouldn't be necessary to manually upload code to the device, the user should be able to select it on the web interface and that's it. Now, if I have a  full TZ list, that's great. But if it changes over time, then eventually it would become unusable and would require the code to be manually updated over USB, which I'd like to avoid. So, I've been also considering letting the user manually define the timezone and the beginning/end of DST if there is any, which wouldn't require any updates as the TZ could be redefined if it changes, but also wouldn't be quite as easy to use as the inclusion of a complete TZ list. The third solution is letting the Pico W automatically update its timezone definition list from somewhere on the internet, but that would need a lot of extra code and conversion and saving them, so I don't know if it would worth all the extra hassle. And if the site where I'm pulling them from goes down, it was all pointless.

Anyway, thanks for all the help, I'll try to make the most of it.
